bundles/org.openhab.core.crypto.cryptostore/src/main/java/org/openhab/core/crypto/cryptostore/CryptoStore.java

```diff
@@ -0,0 +1,328 @@
+
+    public CryptoStore() {
+        try {
+            encryptionCipher = Cipher.getInstance(cipher);
+        } catch (NoSuchAlgorithmException | NoSuchPaddingException e) {
+            logger.debug("Could not get cipher instance", e);
+        }
+    }
+
+    public byte[] generateEncryptionKey() {
+        Key key;
+        try {
+            KeyGenerator keyGenerator = KeyGenerator.getInstance("AES");
+            keyGenerator.init(keySize);
+            key = keyGenerator.generateKey();
+            byte[] newKey = key.getEncoded();
+            this.keyAlgorithm = key.getAlgorithm();
+            return newKey;
+        } catch (NoSuchAlgorithmException e) {
+            logger.debug("Could not generate encryption keys", e);
+        }
+        return new byte[0];
+    }
+
+    private Key convertByteToKey(byte[] keyString) {
+        Key key = new SecretKeySpec(keyString, keyAlgorithm);
+        return key;
+    }
+
+    public String encrypt(String data, Key key) throws Exception {
+        return encrypt(data, key, this.cipher);
+    }
+
+    public String encrypt(String data, Key key, String cipher) throws Exception {
+        byte[] dataInBytes = data.getBytes();
+        Cipher encryptionCipher = this.encryptionCipher;
+        if (encryptionCipher != null) {
+            encryptionCipher.init(Cipher.ENCRYPT_MODE, key);
+            byte[] encryptedBytes = encryptionCipher.doFinal(dataInBytes);
+            return Base64.getEncoder().encodeToString(encryptedBytes);
+        } else {
+            return "";
+        }
+    }
+
+    public String decrypt(String encryptedData, Key key) throws Exception {
+        return decrypt(encryptedData, key, this.cipher);
+    }
+
+    public String decrypt(String encryptedData, Key key, String cipher) throws Exception {
+        byte[] dataInBytes = Base64.getDecoder().decode(encryptedData);
+        Cipher decryptionCipher = Cipher.getInstance(cipher);
+        Cipher encryptionCipher = this.encryptionCipher;
+        if (encryptionCipher != null) {
+            GCMParameterSpec spec = new GCMParameterSpec(keySize, encryptionCipher.getIV());
+            decryptionCipher.init(Cipher.DECRYPT_MODE, key, spec);
+            byte[] decryptedBytes = decryptionCipher.doFinal(dataInBytes);
+            return new String(decryptedBytes);
+        } else {
+            return "";
+        }
+    }
+
+    public void setPrivKey(String privKey, byte[] keyString) throws Exception {
+        Key key = convertByteToKey(keyString);
+        this.privKey = encrypt(privKey, key);
+    }
+
+    public String getPrivKey(byte[] keyString) throws Exception {
+        Key key = convertByteToKey(keyString);
+        return decrypt(this.privKey, key);
+    }
+
+    public void setCert(String cert) {
+        this.cert = cert;
+    }
+
+    public void setCert(Certificate cert) throws CertificateEncodingException {
+        this.cert = new String(Base64.getEncoder().encode(cert.getEncoded()));
```

Comment:
Why is it stored as a String instead of as a Certificate object?

Reply:
I went back and forth on this several times.  It was just easier to store and manipulate as a Base64 encoded String in the end.  If you look through some older versions of the AndroidTV PR I had it in as a Certificate and ended up changing it.

Reply:
What kinds of manipulations would be needed? If a certificate is modified after being signed, the signature (and therefore the entire certificate) becomes invalid.

Reply:
I'll have to go back and look to see what the issue was.  I believe part of it was simply the way exceptions were handled (where Certificate throws several that we would have to catch and deal with, String did not which made the code much cleaner).  Another part of it was that it's very likely that the bindings are dealing with them in base64 (ATV was for example being sent a certificate as part of something else that I had to deal with) so flipping back and forth was becoming complicated.

Reply:
Was mainly a thought because the getter for the certificate returns a Certificate object, so there are more conversions when storing as a string that it would be the other way around. Providing a string as a setter should definitely be supported, but it would probably be better to get an exception when trying to set an invalid string-encoded cert than getting it when retrieving the cert.

Reply:
I went back and looked at the old commits and why I changed this.  This is absolutely because of exception handling.  I was trying to avoid a try/catch inside the class (so that the binding could handle the try/catch and logs were more accurate).  By storing as a string I avoided that (and could simply put it in a throws line for the end user to catch).